[Flang][OpenMP] Add OpenMP standards support doc

kiranchandramohan opened the pull request.

kiranchandramohan committed:
[Flang][OpenMP] Add OpenMP standards support doc

kiranchandramohan commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+| target teams distribute parallel loop construct            | P      | |
+| teams distribute parallel loop simd construct              | P      | |
+| target teams distribute parallel loop simd construct       | P      | |
+
```
I am conservatively marking all target-related constructs (including combined ones) as Partial support. I can correct the entries if they are fully supported.

tblah commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+| distribute parallel loop simd construct                    | P      | |
+| depend clause                                              | P      | Depend clause with array sections are not supported |
+| declare reduction construct                                | N      | |
+| atomic construct extensions                                | Y      | |
```
I don't think there is any support for atomic: seq_cst yet

tblah replied:
`atomic-16.f90` in the gfortran testsuite doesn't parse and looks like it is supposed to work. Surprisingly I don't find any tests for this feature in the Fujitsu testsuite.

kiranchandramohan replied:
The parse error is probably the result of the `end atomic` statement becoming optional for the capture directive in OpenMP 5.0 or later.

tblah commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+## OpenMP 4.0
+
+| Feature                                                    | Status | Comments                                                |
+|------------------------------------------------------------|--------|---------------------------------------------------------|
+| proc_bind clause                                           | Y      | |
```
```suggestion
|------------------------------------------------------------|--------|---------------------------------------------------------|
| Array sections for depend clause                   | N      | |
| proc_bind clause                                           | Y      | |
```

tblah reviewed (commented):
Thanks for putting this together Kiran.

kiranchandramohan reviewed (commented):
Thanks @tblah for the quick review.

tarunprabhu commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+
+This document summarizes OpenMP standards support in Flang. The information is only provided as a guideline. The
+TODOs/Not Yet Implemented messages emitted by the compiler for unimplemented features should be treated as authoritative.
+Standards support is provided upto OpenMP 4.0 for now. It will be extended later for OpenMP 4.5, OpenMP 5.* and OpenMP 6.0.
```
Would it be worth putting in a date here, so there is some indication of how valid this sentence is?

```suggestion
As of March 2025, standards support is provided upto OpenMP 4.0. It will be extended later for OpenMP 4.5, OpenMP 5.* and OpenMP 6.0.
```

Are we working on support for OpenMP > 4.0? If so, perhaps the second sentence could be rephrased to better reflect this. Perhaps

```
We are actively working towards supporting OpenMP 4.5, OpenMP 5.* and OpenMP 6.0. 
```

kiranchandramohan replied:
> Are we working on support for OpenMP > 4.0?

Yes. Some might be working already. Just that I have not spent time characterising the support.

Your suggestions look good. Have made the change.

tarunprabhu reviewed (commented):
Just some minor suggestions

kiranchandramohan committed:
Address review comment

tarunprabhu commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+- **Y** : When the implementation is complete
+- **N** : When the implementation is absent
+
+Note : No distinction is made between the support in the Parser/Semantics, MLIR or Lowering support, and OpenMPIRBuilder support.
```
Another optional rephrase (if I understand your intent correctly)

```suggestion
Note : No distinction is made between the support in Parser/Semantics, MLIR, Lowering or the OpenMPIRBuilder.
```

kiranchandramohan replied:
Done

kiranchandramohan committed:
Address review comments

NimishMishra reviewed (commented):
Thanks for compiling this. I am okay with the list and their current status.

Are we not adding `task` construct though?

kiranktp commented:
I think we are missing taskgroup as well.

kiranchandramohan commented:
> Are we not adding task construct though?

> I think we are missing taskgroup as well.

These came in OpenMP 3.0/3.1. Or do you mean there is something unsupported that you want to report?

NimishMishra commented:
> These came in OpenMP 3.0/3.1. Or do you mean there is something unsupported that you want to report?

Okay understood. I think we want to say we support OpenMP 4.0 + additional features which we list in the table.  I got confused with the line _"The standards support information is provided as a table with three columns that are self explanatory. The Status column uses the letters **P**, **Y**, **N** for the implementation status"_

mjklemm commented on flang/docs/OpenMPStandardsSupport.md:
````diff
@@ -0,0 +1,62 @@
+---
+```
+
+This document summarizes OpenMP standards support in Flang. The information is only provided as a guideline. The
````
```suggestion
This document summarizes the supported features of the OpenMP API in Flang. The information is only provided as a guideline. The
```

(OpenMP is not a standard in the true sense, like with the ISO standards.)

kiranchandramohan replied:
Would it be OK to go with OpenMP specification?
`This document summarizes the supported features of the OpenMP specification in Flang`

kiranchandramohan replied:
Done.

mjklemm commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+
+This document summarizes OpenMP standards support in Flang. The information is only provided as a guideline. The
+TODOs/Not Yet Implemented messages emitted by the compiler for unimplemented features should be treated as authoritative.
+As of March 2025, standards support is provided upto OpenMP 4.0. We are actively working towards supporting OpenMP 4.5, OpenMP 5.* and OpenMP 6.0.
```
I know that it has been requested to have a date in the document, but I would treat the modification date of the file in the commit log as being more accurate than the date in the document that could easily be missed when updating the file (plus, it does not protect us from the document becoming outdated over time).

h-vetinari replied:
Note that https://clang.llvm.org/docs/OpenMPSupport.html has split OpenMP support into C/C++ & Fortran already. Presumably it makes sense to either fully join those documents, or remove the Fortran side from the clang docs, and just add a reference to this document.

Also typo
```suggestion
As of March 2025, standards support is provided up to OpenMP 4.0. We are actively working towards supporting OpenMP 4.5, OpenMP 5.* and OpenMP 6.0.
```
From the table below, OpenMP 4.0 support is not 100% complete yet, so maybe "up to" gives a wrong impression here; I'd read it as "4.0 support is fully implemented".

kiranchandramohan replied:
> Note that https://clang.llvm.org/docs/OpenMPSupport.html has split OpenMP support into C/C++ & Fortran already. Presumably it makes sense to either fully join those documents, or remove the Fortran side from the clang docs, and just add a reference to this document.

Would it be all right to address this separately? Ideally, each Frontend should list its OpenMP support separately. Removing the entry from Clang would require a separate set of reviewers.

kiranchandramohan replied:
Rewrote the statement regarding standards support.

Changing the support statement in Clang docs is out of scope for this PR.

mjklemm commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+| target teams distribute parallel loop simd construct       | P      | |
+
+## OpenMP 3.1, OpenMP 2.5, OpenMP 1.1
+All features except a few corner cases in atomic, copyin constructs/clauses are supported 
```
Should these corner cases be made explicit?

kiranchandramohan replied:
Done.

h-vetinari commented on flang/docs/OpenMPStandardsSupport.md:
```diff
@@ -0,0 +1,62 @@
+
+The standards support information is provided as a table with three columns that are self explanatory. The Status column uses
+the letters **P**, **Y**, **N** for the implementation status:
+- **P** : When the implementation is incomplete for a few cases
+- **Y** : When the implementation is complete
+- **N** : When the implementation is absent
```
Might want to explain that this is Yes/No/Partial (presumably)

kiranchandramohan replied:
Done

h-vetinari reviewed (commented):
I'm excited to see this, thanks for adding it! This will close an old issue of mine :)
- #62310

kiranchandramohan committed:
Address review comments

OpenMP standard -> OpenMP API
specify corner cases for atomic, threadprivate
Nit spellings

kiranchandramohan committed:
Rename file in index.md

bhandarkar-pranav commented on flang/docs/OpenMPSupport.md:
```diff
@@ -0,0 +1,64 @@
+
+This document outlines the OpenMP API features supported by Flang. It is intended as a general reference.
+For the most accurate information on unimplemented features, rely on the compiler’s TODO or “Not Yet Implemented”
+messages, which are considered authoritative. Flang provides full support for OpenMP 3.1 and partial support for
```
Nit: Like you mention in the section on previous standards (<=3.1) there are some, albeit very few, corner cases that are not supported. I think this line here is then misleading; it'll make a reader believe 3.1 is fully supported and chances are he won't bother to scroll down to the section below to find info about the corner cases. Perhaps consider something like, but not necessary exactly, the following

```
Aside from very few corner cases, Flang provides full support for OpenMP 3.1 [See detail here](#OpenMP 3.1,-OpenMP 2.5,-OpenMP-1.1). There is also partial (in development) support for OpenMP 4.0
```

(Note: I may have got the link syntax wrong especially around the use of commas).

kiranchandramohan replied:
Done. Thanks.

bhandarkar-pranav reviewed (approved):
Thank you, Kiran. Just one minor nit.

kiranchandramohan committed:
Address comments from Pranav

kiranchandramohan committed:
Fix link - attempt 1

kiranchandramohan committed:
Remove redundant sentence

kparzysz reviewed (approved):
LGTM.

For any `target` construct we don't support the `device` clause.  For `distribute` we don't support `dist_schedule`.  For `teams`, we don't fully support `reduction` (e.g. no `declare reduction`).

kiranchandramohan committed:
Update target directives using info provided by Kryztof

kiranchandramohan commented:
I plan to submit this PR early next week if there are no objections.

NimishMishra reviewed (approved):
LGTM. Thanks for this.
